syntax = "proto3";
option go_package = "kademlia/service";

service KademliaService {
        rpc PING (PingCheck) returns (NodeInfo) {}
        rpc STORE (StoreRequest) returns (StoreResult) {}
        rpc FIND_NODE (LookupRequest) returns (LookupResponse) {}
        rpc FIND_VALUE (Key) returns (Value) {}
}

message PingCheck {
}

message StoreRequest {
        int32 key = 1;
        int32 value = 2;
}

message StoreResult {
        bool result = 1;
        string message = 2;
}

message LookupRequest {
        optional NodeInfo requester = 1;
        string targetNodeId = 2;
        optional uint64 magicCookie = 3;
}

message NodeInfo {
        string ip = 1;
        int32 port = 2;
        string id = 3;
}

message NodeInfoLookup {
        NodeInfo nodeDetails = 1;
        uint32 distanceToTarget = 2;
}

message LookupResponse {
        repeated NodeInfoLookup nodes = 1;
}

message Key {
        int32 key = 1;
}

message Value {
        int32 value = 1;
}